File: R package/src/kronecker.c
/* $ID: kronecker.c, last updated 2020-09-21, F.Osorio */

#include "fastmatrix.h"

void
kronecker_prod(double *a, int *arow, int *acol, double *b, int *brow, int *bcol, double *prod)
{ /* kronecker product between two matrices */
  int nrow = *arow * *brow, irow, jcol;
  double z;

  for (int i = 0; i < *arow; i++) {
    irow = i * *brow;
    for (int j = 0; j < *acol; j++) {
      jcol = j * *bcol;
      z = a[j * *arow + i];
      for (int k = 0; k < *brow; k++) {
        for (int l = 0; l < *bcol; l++)
          prod[(jcol + l) * nrow + irow + k] = z * b[l * *brow + k];
      }
    }
  }
}

File: R package/R/kronecker.R
## ID: kronecker.R, last updated 2020-08-16, F.Osorio

kronecker.prod <- function(x, y = x)
{ ## returns the kronecker product of two matrices
  if (is.data.frame(x))
    x <- as.matrix(x)
  if (is.vector(x))
    x <- as.matrix(x)
  if (!is.matrix(x))
    stop("supply a matrix-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric matrix" )

  if (is.data.frame(y))
    y <- as.matrix(y)
  if (is.vector(y))
    y <- as.matrix(y)
  if (!is.matrix(y))
    stop("supply a matrix-like 'y'")
  if (!is.numeric(y))
    stop("argument y is not a numeric matrix" )

  dx <- dim(x)
  dy <- dim(y)
  z <- double(prod(dx * dy))
  dim(z) <- dx * dy

  storage.mode(x) <- "double"
  storage.mode(y) <- "double"
  storage.mode(z) <- "double"

  z <- .C("kronecker_prod",
          x = x,
          xrow = as.integer(dx[1]),
          xcol = as.integer(dx[2]),
          y = y,
          yrow = as.integer(dy[1]),
          ycol = as.integer(dy[2]),
          z = z)$z
  z
}
